feature/ocl: Improve Linux CL/GL sharing support

smunaut opened the pull request:
This PR is aimed at drastically improving the support for the CL/GL sharing extension on linux.
The current support is not really usable as it only supports a few texture format, and only on EGL contexts. It is also pretty buggy since it requires the texture to be bound when placing the CL call to share it which is just plain wrong and will not work in many applications.
This new version makes used of the "official" interop extension from MESA which is available for GLX and EGL contexts, allows sharing of buffers and not just texture and supports many more formats. 
This is still far from being a fully compliant / full featured version of the extension, but it's a big step forward in my opinion and allows to run some real applications. I've tested gr-fosphor (SDR spectrum display) and Davinci Resolve as examples. Both of theses don't work without theses improvements.

Resolves: #659
Resolves: #667 

Signed-off-by: Sylvain Munaut <tnt@246tNt.com>

k1gen commented:
I have tested this on my machine, and the app I wanted to work (DaVinci Resolve) - worked flawlessly on UHD620!

smunaut commented:
So any chance someone from intel can review / comment or just general advise on what is required to get this merged ?

JablonskiMateusz commented on opencl/source/sharings/gl/linux/gl_buffer_linux.cpp:
```diff
@@ -24,15 +24,44 @@ using namespace NEO;
-    bufferInfo.bufferName = bufferId;
 
+    /* Prepare export request */
+    struct mesa_glinterop_export_in tex_in = {};
+    struct mesa_glinterop_export_out tex_out = {};
```
please use camelCase

JablonskiMateusz commented on opencl/source/sharings/gl/linux/gl_buffer_linux.cpp:
```diff
@@ -24,15 +24,44 @@ using namespace NEO;
     }
 
+    /* Map result for rest of the function */
+    CL_GL_BUFFER_INFO bufferInfo = {
+        .bufferName        = bufferId,
+        .globalShareHandle = (unsigned int) tex_out.dmabuf_fd,
+        .bufferSize        = (GLint) tex_out.buf_size,
+        .bufferOffset      = (GLint) tex_out.buf_offset,
+    };
```
please don't use C-style casts

JablonskiMateusz commented on opencl/source/sharings/gl/linux/gl_buffer_linux.cpp:
```diff
@@ -168,14 +199,20 @@ GraphicsAllocation *GlBuffer::createGraphicsAllocation(Context *context, unsigne
+        else
+        {
+            auto helper = context->getDevice(0)->getRootDeviceEnvironment().getGmmHelper();
+            StorageInfo si = {};
```
```suggestion
            StorageInfo storageInfo = {};
```

please use meaningful name

JablonskiMateusz commented on opencl/source/sharings/gl/linux/gl_context_guard_linux.cpp:
```diff
@@ -10,6 +10,7 @@
 
 namespace NEO {
 GLContextGuard::GLContextGuard(GLSharingFunctions &sharingFcns) : sharingFunctions(&sharingFcns) {
+#if 0
```
if code is not needed then remove it

JablonskiMateusz commented on opencl/source/sharings/gl/linux/gl_context_guard_linux.cpp:
```diff
@@ -27,12 +28,15 @@ GLContextGuard::GLContextGuard(GLSharingFunctions &sharingFcns) : sharingFunctio
 }
 
 GLContextGuard::~GLContextGuard() {
+#if 0
```
if code is not needed then remove it

JablonskiMateusz commented on opencl/source/sharings/gl/linux/gl_sync_event_linux.cpp:
```diff
@@ -27,10 +27,13 @@ GlSyncEvent::GlSyncEvent(Context &context, const GL_CL_SYNC_INFO &sync)
 }
 
 GlSyncEvent::~GlSyncEvent() {
+#if 0
```
same in this file

smunaut replied:
Although here we should have something to properly support sync point, the current code in there doesn't work (just crash calling to nullptr) so it's removed until someone implements it properly.

JablonskiMateusz commented on opencl/source/sharings/gl/linux/gl_texture_linux.cpp:
```diff
@@ -39,52 +39,61 @@ Image *GlTexture::createSharedGlTexture(Context *context, cl_mem_flags flags, cl
-    texInfo.target = getBaseTargetType(target);
-    if (texInfo.target != GL_TEXTURE_2D) {
+    /* Prepare export request */
+    struct mesa_glinterop_export_in tex_in = {};
+    struct mesa_glinterop_export_out tex_out = {};
```
please use camelCase

JablonskiMateusz commented on opencl/source/sharings/gl/linux/gl_texture_linux.cpp:
```diff
@@ -39,52 +39,61 @@ Image *GlTexture::createSharedGlTexture(Context *context, cl_mem_flags flags, cl
-    }
+    /* Map result for rest of the function */
+    CL_GL_RESOURCE_INFO texInfo = {
+        .name                = tex_in.obj,
+        .globalShareHandle   = (unsigned int) tex_out.dmabuf_fd,
+        .glInternalFormat    = (GLint) tex_out.internal_format,
+        .textureBufferSize   = (GLint) tex_out.buf_size,
+        .textureBufferOffset = (GLint) tex_out.buf_offset,
```
please don't use C-style casts

JablonskiMateusz commented on opencl/source/sharings/gl/linux/gl_texture_linux.cpp:
```diff
@@ -177,20 +184,23 @@ Image *GlTexture::createSharedGlTexture(Context *context, cl_mem_flags flags, cl
-        }
+
+    /* Prepare flush request */
+    struct mesa_glinterop_export_in tex_in = {};
```
please use camelCase

JablonskiMateusz commented on opencl/source/sharings/gl/linux/gl_texture_linux.cpp:
```diff
@@ -177,20 +184,23 @@ Image *GlTexture::createSharedGlTexture(Context *context, cl_mem_flags flags, cl
+    tex_in.miplevel = this->miplevel;
+
+    /* Call MESA interop */
+    int rv = sharingFunctions->flushObjects(1, &tex_in, nullptr);
```
please use meaningful name

JablonskiMateusz commented on third_party/opengl_headers/GL/mesa_glinterop.h:
```diff
@@ -0,0 +1,366 @@
+/*
```
please mention source of the file in commit msg

smunaut commented on opencl/source/sharings/gl/linux/gl_sharing_linux.cpp:
```diff
@@ -11,41 +11,16 @@
-    glDeviceHandle = contextInfo.deviceHandle;
-
-    return retVal;
+    std::unique_ptr<OsLibrary> dynLibrary(OsLibrary::load(""));;
```
So there are several things :

(1) We can't link directly to any GL library. We have to use which ever one the application is using and it might be a special one for whatever reason, so we can't add `-lGL` to the Makefile or load any library by name explicitely.

(2) We can't just "use" GL symbol without linking and assume they will be present because if we do, then any application that doesn't use GL interop and doesn't use GL at all will end up with a "symbol not found" error.

(3) But we know that _if_ the user want to use GL interop, then at somepoint, it will have to have loaded a GL library in the global scope which mean at that point we can query it to find the few functions we need. And if that lookup fails, we know the user will not need GL interop.

JablonskiMateusz replied:
(1) - I know that we cannot link using load-time linking as GL sharing is optional but what is an issue with run-time linking by name? When user is using GL interop then GL library need to be available in the system
(2) I totally agree, we cannot use GL symbols directly 
(3) This approach assumes that GL is loaded before OCL is loaded but I can imagine a scenario where CL is initialized before GL is even loaded and in such case I hope should try to load GL to determine if sharing is available.

smunaut commented:
Thanks for the review, I'll try to address these next week.

smunaut commented:
@JablonskiMateusz  Ok, so it took me a while longer to finally get to it, but I have addressed the comments above, rebased over master (and also tested only the 23.30 release) and seems to all be fine.

JablonskiMateusz commented on opencl/source/sharings/gl/linux/gl_context_guard_linux.cpp:
```diff
@@ -10,29 +10,8 @@
 
 namespace NEO {
 GLContextGuard::GLContextGuard(GLSharingFunctions &sharingFcns) : sharingFunctions(&sharingFcns) {
```
is this class still needed ?

smunaut replied:
Yes, it's used by the "common code" shared between linux / windows. The alternative would be `#ifdef` in the common code but I wanted to avoid touching windows / common code as much as possible.

JablonskiMateusz commented on opencl/source/sharings/gl/linux/gl_buffer_linux.cpp:
```diff
@@ -24,15 +24,60 @@ using namespace NEO;
+    struct mesa_glinterop_export_in objIn = {};
+    struct mesa_glinterop_export_out objOut = {};
+
+    objIn.version  = MESA_GLINTEROP_EXPORT_IN_VERSION;
+    objIn.target   = GL_ARRAY_BUFFER;
+    objIn.obj      = bufferId;
```
please format all files using clang-format 15

smunaut replied:
Ack, will do.

smunaut commented:
@JablonskiMateusz  Responding here for the libGL thing because for some reason, github doesn't me put a reply to that specific comment :shrug: 

(1) Because for whatever reason the user is free to have loaded (through whatever mechanism) any GL library and possibly not the first one that would be returned by a normal libdl lookup of `libGL.so.0`. Then there is also the EGL/GLX thing where we have no idea a-priori which one the user will want to use and we don't want to force loading the wrong one in the same address space, so it's easier to lookup what the user chose by looking at already available symbols.

(3) In the `initGLFunctions` it's not an issue because this is called when creating the CL context but at that point the user must have given a reference to the current GL context in the parameter list which means that GL must have been loaded already.

In the `isEnabled` case (to decide wether we show the extension or not), it's less clear. All the functionality of the extensions depend on a GL context being present first so to me it make sense to advertise it only if GL is loaded. But we could just as well just return `true` all the time and leave it up to the user to actually make sure it gets a valid GL context in time before trying to actually make use of it.

smunaut commented:
But we don't know which GL library ... it's not our choice.

It's be 100% valid for a user to do `dlopen("/path/to/some/custom/libGL.so.1234", RTLD_GLOBAL);`  to load a GL library that's not the system default one ... So we can't hard code any names or path anywhere.

If you don't want to assume GL is already loaded to return the extension, then I would return it all the time. Which from reading the extension spec is perfectly valid since all that's needed is that there is _some_ way/combination of GL context/device for which we could support it.

eero-t commented:
Btw. @smunaut have you looked (e.g. latrace'd) what Nvidia and AMD OpenCL drivers do?

smunaut commented:
@eero-t 

AMD looksup symbols in the global scope without loading anything it self, you can see it here : 

https://github.com/ROCm-Developer-Tools/clr/blob/5914ac3c6e9b3848023a7fa25e19e560b1c38541/rocclr/device/rocm/rocglinterop.cpp#L71C38-L71C69

(Now, here they lookup `MesaGLInteropGLXExportObject` directly which is "the old way" of doing it because until recently that was the only way since glxGetProcAddress wouldn't support that symbol but it was shown that this direct method didn't work when using EGL + GLVND so mesa added those interop symbols to the normal GetProcAddress ... but it's the same idea, they search in the global scope what's already there and don't load anything themselves).

The AMD implementation always advertises the extension, it's statically present in the list of supported extension and it doesn't do any checks before it returns it AFAICT. ( https://github.com/ROCm-Developer-Tools/clr/blob/5914ac3c6e9b3848023a7fa25e19e560b1c38541/rocclr/device/device.hpp#L204 )

Same in Mesa's rusticl : https://gitlab.freedesktop.org/mesa/mesa/-/commit/02f7d3640021f23d54022c6fbdf1304c44672033#cb331a47f57adfd7e2565e461a25af9ec00ff8e6_0_32
They call `dlsym` with a `NULL` first argument to just lookup in what's already loaded rather than any form of explicit linking.
It will also not return the extension if GL is not loaded already.

JablonskiMateusz commented:
I think we may self load in this case. Extension can be exposed always but to work with sharing GL library needs to be loaded.

JablonskiMateusz commented on opencl/source/sharings/gl/linux/lin_enable_gl.cpp:
```diff
@@ -76,17 +81,19 @@ void GlSharingBuilderFactory::fillGlobalDispatchTable() {
-               "cl_khr_gl_event "
-               "cl_khr_gl_msaa_sharing ";
-    } else if (GLSharingFunctionsLinux::isGlSharingEnabled()) {
+    auto isGlSharingEnabled = GLSharingFunctionsLinux::isGlSharingEnabled();
```
based on our discussion. It should be not depend on gl functions availability

smunaut commented:
* Rebased to master
* Removed the isGLSharingEnabled check and always advertise extension
* Adapted to the final version of the mesa interop interface that is now part of git master on the mesa repo.

JablonskiMateusz commented:
@smunaut  please adjust commit msg to https://github.com/intel/compute-runtime/blob/master/CONTRIBUTING.md#commit-message-guidelines I think squash to single commit would be needed

smunaut commented:
Is squashing really needed ?   It makes the diff really hard to read and also means I can't put explanation of why a particular change was done that way in the commit message since it's all globbed up into a giant unreadable diff.

All the commits would belong in `feature/ocl` AFAICT.

JablonskiMateusz commented:
it will be squashed in the end but we can do this on our side. Let's keep it here without squashing. However, I got some errors when applying the PR:

> ../../../opencl/source/sharings/gl/linux/gl_buffer_linux.cpp:225:51: error: no matching constructor for initialization of 'NEO::Gmm'
            graphicsAllocation->setDefaultGmm(new Gmm(helper,
                                                  ^   ~~~~~~~
../../../shared/source/gmm_helper/gmm.h:45:5: note: candidate constructor not viable: requires 7 arguments, but 8 were provided
    Gmm(GmmHelper *gmmHelper, const void *alignedPtr, size_t alignedSize, size_t alignment,
    ^
../../../shared/source/gmm_helper/gmm.h:44:5: note: candidate constructor not viable: requires 4 arguments, but 8 were provided
    Gmm(GmmHelper *gmmHelper, ImageInfo &inputOutputImgInfo, const StorageInfo &storageInfo, bool preferCompressed);
    ^
../../../shared/source/gmm_helper/gmm.h:48:5: note: candidate constructor not viable: requires 3 arguments, but 8 were provided
    Gmm(GmmHelper *gmmHelper, GMM_RESOURCE_INFO *inputGmm, bool openingHandle);
    ^
../../../shared/source/gmm_helper/gmm.h:47:5: note: candidate constructor not viable: requires 2 arguments, but 8 were provided
    Gmm(GmmHelper *gmmHelper, GMM_RESOURCE_INFO *inputGmm);
    ^
../../../shared/source/gmm_helper/gmm.h:40:7: note: candidate constructor (the implicit copy constructor) not viable: requires 1 argument, but 8 were provided
class Gmm {
      ^
../../../shared/source/gmm_helper/gmm.h:43:5: note: candidate constructor not viable: requires 0 arguments, but 8 were provided
    Gmm() = delete;
    ^
1 error generated.

Also, when fixed the issue locally I got test failures:

>  97x TEST_F(GlSharingTextureTests, givenMockGlWhen2dGlTextureIsCreatedThenMemObjectHasGlHandler) {
 98x     cl_int retVal = CL_INVALID_VALUE;
 99x
100x     glSharing->uploadDataToTextureInfo(textureId);
101x
102t>    auto glTexture = GlTexture::createSharedGlTexture(clContext.get(), (cl_mem_flags)0, GL_TEXTURE_2D, 0, textureId, &retVal);
103x     ASSERT_NE(nullptr, glTexture);

>  46x     texIn.version = 2;
 47x     texIn.target = getBaseTargetType(target);
 48x     texIn.obj = texture;
 49x     texIn.miplevel = miplevel;
 50x
 51x     switch (flags) {
 52x     case CL_MEM_READ_ONLY:
 53x         texIn.access = MESA_GLINTEROP_ACCESS_READ_ONLY;
 54x         break;
 55x     case CL_MEM_WRITE_ONLY:
 56x         texIn.access = MESA_GLINTEROP_ACCESS_WRITE_ONLY;
 57x         break;
 58x     case CL_MEM_READ_WRITE:
 59x         texIn.access = MESA_GLINTEROP_ACCESS_READ_WRITE;
 60x         break;
 61x     default:
 62t>        errorCode.set(CL_INVALID_VALUE);
 63x         return nullptr;
 64x     }


Could you please ensure that this code is working after rebase?

smunaut commented:
I'm doing what I can to test but again, I'm trying to rebase to master and build here but I just can't ... `master` just doesn't build and it ends with `ocloc` segfaulting currently, so there isn't much testing I can do.

I can build `23.48.27912.11`  so currently that's the most recent thing I could rebase and test onto. Both `23.52.28202.13` and `master` don't build for me.

smunaut commented:
Ok, I managed to build master on my machine (seems there is a requirement for some specific version of dependencies that are not checked by the build system).

So I rebased and checked it build. I added a couple of small fixes and removed non-applicable broken tests.

Running `make run_adlp_0_ocl_tests` passes without error.  I can't run all the tests because some of them fail on my machine even without this patch and the testing stops at the first failure ... But this specific test target should check everything that is affected by this patch set.

I didn't change the commit message since if you squash on your side you'll have to make a new one anyway so at that point you can use the `feature/ocl` prefix for it at that time.

eero-t commented:
What were the specific dependency versions that build system did not check?

(I normally take latest tagged releases of GMMlib, SPIRV-*, vc-intrinsics, IGC, L0, compute-runtime, and build them with Ubuntu using system LLVM 14 packages: clang, libopencl-clang, llvm-spirv, libllvmspirvlib, and haven't had any problems.  I haven't tried building `master` though...)

smunaut commented:
I was trying to build both `master` and `23.52.28202.13` while having :
* `intel-graphics-compiler-1.0.15368.3` 
* `intel-vc-intrinsics-0.13.0`

And `cmake` accepted those fine and it started building but it ended with `ocloc` segfaulting.

After I upgraded to :
* `intel-graphics-compiler-1.0.15770.7`
* `dev-libs/intel-vc-intrinsics-0.16.0`

then it worked and I was able to build.

smunaut commented:
Just rebased on top of master again.
Checked that it builds, works in a coupld of applications, and that `make run_adlp_0_ocl_tests` works too.

JablonskiMateusz commented:
Hi @smunaut 
I'll try merge it but I need additional input from your side.
Please update PR title and PR description to meet our commit msg restrictions https://github.com/intel/compute-runtime/blob/master/CONTRIBUTING.md#commit-message-guidelines

smunaut commented:
@JablonskiMateusz I've updated the PR title and description to match those guidelines.  I didn't update the commits themselves since AFAIR you said you would squash them all together and I assume the final commit title/description will be taken from the PR.

smunaut commented:
Merged outside of github.

Thanks @JablonskiMateusz
